### PR TITLE
feat(UI): Fire control control changes

### LINKED ISSUE #4714: Optional individual turret tracking setting for each turret type
**Is your feature request related to a problem? Please describe.**

After turret tracking nerf some turrets turn too slow to be effective in opportunistic mode and require focused mode to be effective - while others remain perfectly viable in opportunistic mode. 

On flagships with mostly turret hardpoints, having different types of turrets e.g. 2x fast turrets with mediocre damage for fast targets and 2x slow but powerful turrets for main target - to compensate for lack of gun hardpoints - is essential, but on either setting one of the turret flavors - and overall ship efficiency - suffers greatly. 

The situation is particularly dramatic for Korath Disruptor turret, making it pretty much useless in opportunistic mode for solo players or small fleets - when player's flagship usually needs protection from drones and fighters, thus fast turrets in opportunistic mode are a necessity. 

Turrets that particularly benefit from focused mode:
* Dual Sunbeam Turret
* Electron Turret
* Korath Disruptor
* Plasma Turret
* Sunbeam Turret

Turrets that particularly benefit from opportunistic mode:
* Heavy Laser Turret 
* Korath Repeater Turret
* Laser Turret
* Moonbeam Turret
* Proton Turret
* Pug Gridfire Turret
* Pug Zapper Turret
* Pulse Turret
* Quad Blaster Turret
* Quarg Skylance
* Thrasher Turret

Turrets that benefit from either mode depending on the situation:
* Bombardment Turret
* Heliarch Attractor
* Heliarch Repulsor
* Korath Banisher
* Korath Grab-Strike
* Korath Slicer Turret

**Describe the solution you'd like**

I propose "custom" turret tracking setting activating advanced menu with individual turret tracking setting for each turret type. 

**Additional context**

This change will not affect overall balance of the game in any noticeable way, but will help and improve usable outfits diversity for players who fly solo or with fleets small enough that more effective turret control on flagship can make any impact. Particularly, this feature would make Korath Disruptor less useless in described circumstances, and make turrets with slow tracking, and consequently ships with mostly turrets hardpoints, less undesirable compared to their gun counterparts.

### PR DESCRIPTION
**Pull request by warp-core**
Builds on #6730. Should close #4714.

## Feature Details
Adds a new panel, accessible via a button on the ship info panel, that will allow the player to set defensive and opportunistic states on their weapons as well as displaying some relevant information.

It's not done yet, as you can see:

## UI Screenshots
image

## Testing Done
{{describe how you tested the new feature}}

## Performance Impact
{{describe any performance impact (positive or negative). "N/A" if no performance-critical code is changed. }}

### CODE REVIEW

**Comment by Zitchas** [comment]
Oh, excellent! Thanks for working on this; I'm very much looking forward to being able to set different kinds of behavior to different turrets.

**Comment by warp-core** [comment]
From #4714, @Zitchas suggested a slider that the game uses to automatically assign opportunistic to turrets based on  tracking speed, I think this can be added to this panel alongside the per hardpoint option already available (to make setting it on many turrets across many ships faster).
There were mentions of tooltips in the ship info panel on hovering over each weapon with icons to indicate opportunisticness, I also think it'd be good to add icons for defensiveness, always/frugal/never ammo expenditure and if the weapon is disabled, though that might be better for #6730 since it's not specific to the new panel.

**Comment by warp-core** [comment]
What's the most commits a single PR has had made to its branch here?

**Comment by Zitchas** [comment]
As MCO pointed out, PR #3761 had 188 commits and 1296 items in its conversation; and #4871 had 252 commits (but only 8 items of conversation). Can we tell which one I worked on the longest as a github branch? lol.

**Comment by warp-core** [comment]
image
image

**Comment by warp-core** [comment]
That was... interesting. Hope it didn't break anything too badly.

**Comment by warp-core** [comment]
image

**Review by samrocketman — COMMENTED** [review]
I posted a video in discord https://discord.com/channels/251118043411775489/947227869036286033/978472385323880508

Here's my review so far:

> As a player, when I reorganize weapons I expect all of the things associated with said weapon to move with it.  In this case, I expected the settings set for that turret to transfer when I moved the weapon.  Instead, the hardpoint settings do not transfer.  Now arguably I understand these settings are for the hardpoint itself.  However, as a player (not thinking in code) to me intuitively I'm setting those settings for the weapon.  I would expect the settings to transfer when I reorganize the weapons.

**Comment by warp-core** [comment]
@samrocketman swapping hardpoints should not take the fire control settings with the hardpoint.

**Comment by Saturnalia9** [comment]
Is it currently possible to turn a weapon off from manual fire? Such as how PD/AM turrets work, just with defensive turrets instead?

**Comment by warp-core** [comment]
> Is it currently possible to turn a weapon off from manual fire? Such as how PD/AM turrets work, just with defensive turrets instead?

Not at the moment. I can add a 'manual control' option.

**Comment by warp-core** [comment]
There's currently a bug in this where some (but not necessarily all) gun and turret hardpoint locations get set close to (but not exactly) (0,0). Not sure what's causing it.
Thanks to @Saturnalia9 for first telling me about this.

File: source/WeaponConfigPanel.cpp [hunk 1]
```diff
@@ -0,0 +1,548 @@
+/*WeaponConfigPanel.cpp
+Copyright (c) 2022 by warpcore
+
+Endless Sky is free software: you can redistribute it and/or modify it under the
+terms of the GNU General Public License as published by the Free Software
+Foundation, either version 3 of the License, or (at your option) any later version.
+
+Endless Sky is distributed in the hope that it will be useful, but WITHOUT ANY
+WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
+PARTICULAR PURPOSE.  See the GNU General Public License for more details.
```
Comment (tibetiroka) [inline source/WeaponConfigPanel.cpp:10]:
```suggestion
PARTICULAR PURPOSE. See the GNU General Public License for more details.

You should have received a copy of the GNU General Public License along with
this program. If not, see <https://www.gnu.org/licenses/>.
```
The standard copyright header was changed recently.

File: source/WeaponConfigPanel.h [hunk 2]
```diff
@@ -0,0 +1,96 @@
+/*WeaponConfigPanel.h
+Copyright (c) 2022 by warpcore
+
+Endless Sky is free software: you can redistribute it and/or modify it under the
+terms of the GNU General Public License as published by the Free Software
+Foundation, either version 3 of the License, or (at your option) any later version.
+
+Endless Sky is distributed in the hope that it will be useful, but WITHOUT ANY
+WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
+PARTICULAR PURPOSE.  See the GNU General Public License for more details.
```
Comment (tibetiroka) [inline source/WeaponConfigPanel.h:10]:
```suggestion
PARTICULAR PURPOSE. See the GNU General Public License for more details.

You should have received a copy of the GNU General Public License along with
this program. If not, see <https://www.gnu.org/licenses/>.
```

**Review by tibetiroka — CHANGES_REQUESTED** [review]

**Comment by Saturnalia9** [comment]
> There's currently a bug in this where some (but not necessarily all) gun and turret hardpoint locations get set close to (but not exactly) (0,0). Not sure what's causing it. Thanks to @Saturnalia9 for first telling me about this.

Did this ever get fixed?

**Comment by Amazinite** [comment]
Closing this PR since it has been in the draft phase for some time. You can reopen it once you've finalized your changes and it's ready for review.

**Pull request closed without merging**